seismicpro/survey/headers_checks.py

```diff
@@ -4,119 +4,159 @@
 from textwrap import wrap
 
 import numpy as np
+import polars as pl
 from sklearn.neighbors import RadiusNeighborsRegressor
 
-from ..utils import to_list
+from seismicpro.utils import to_list
 
 
-def warn_list(title, warning_list, width=80):
-    """Warn about several issues listed in `warning_list`."""
+def format_warnings(title, warning_list, width=80):
+    """Format a list of warnings into a single string."""
     n_warnings = len(warning_list)
     if n_warnings == 0:
-        return
+        return None
     ix_len = len(str(n_warnings))
     wrap_space = 3 + ix_len
     wrap_sep = "\n" + " " * wrap_space
-    line_sep = "\n\n" + "-" * width
     warning_list = [wrap_sep.join(wrap(warn_str, width=width-wrap_space)) for warn_str in warning_list]
-    warning_msg = line_sep + "\n\n" + "\n".join(wrap(title, width=width))
-    warning_msg += "".join([f"\n\n {i+1:{ix_len}d}. {warn_str}" for i, warn_str in enumerate(warning_list)]) + line_sep
-    warnings.warn(warning_msg)
+    warning_msg = "\n".join(wrap(title, width=width))
+    warning_msg += "".join([f"\n\n {i+1:{ix_len}d}. {warn_str}" for i, warn_str in enumerate(warning_list)])
+    return warning_msg
+
+
+def format_warning(warning_str, width=80):
+    """Format a warning string by adding a separator before and after it."""
+    warning_sep = "-" * width
+    return "".join(["\n\n", warning_sep, "\n\n", warning_str, "\n\n", warning_sep])
+
+
+def isclose_polars(expr1, expr2, rtol=1e-5, atol=1e-8):
+    """Return whether two `polars` expressions are close within given tolerances."""
+    return (expr1 - expr2).abs() <= atol + rtol * expr2.abs()
 
 
 # pylint: disable-next=too-many-statements
-def validate_trace_headers(headers, offset_atol=10, cdp_atol=10, elevation_atol=5, elevation_radius=50):
-    """Validate trace headers for consistency. `headers` `DataFrame` is expected to have reset index."""
-    n_traces = len(headers)
-    if n_traces == 0:
-        return
+def _validate_trace_headers(headers, offset_atol=10, cdp_atol=10, elevation_atol=5, elevation_radius=50, width=80):
+    """Validate trace headers for consistency and return either a string with found problems or `None` if all checks
+    have successfully passed or no checks can been performed. `headers` is expected to be a `polars.DataFrame`."""
+    if headers.is_empty():
+        return None
 
-    msg_list = []
+    n_traces = len(headers)
+    loaded_columns = set(headers.columns)
+    empty_mask = headers.select((pl.col("*") == 0).all()).row(0)
+    empty_columns = {col for col, mask in zip(headers.columns, empty_mask) if mask}
```

Comment:
To explicitely secure order
```suggestion
    empty_mask = headers.select((pl.col("*") == 0).all()).row(0, named=True)
    empty_columns = {col for col, mask in empty_mask.items() if mask}
```

Reply:
Fixed both here and in `_validate_source_headers` and `_validate_receiver_headers`